File: algorithms/data-structures/linked-list-operations/python/reverse_linked_list.py
class _Node:
    __slots__ = ("value", "next")

    def __init__(self, value: int) -> None:
        self.value = value
        self.next: "_Node | None" = None


def _build_list(arr: list[int]) -> "_Node | None":
    if not arr:
        return None
    head = _Node(arr[0])
    current = head
    for val in arr[1:]:
        current.next = _Node(val)
        current = current.next
    return head


def _to_array(head: "_Node | None") -> list[int]:
    result: list[int] = []
    current = head
    while current is not None:
        result.append(current.value)
        current = current.next
    return result


def reverse_linked_list(arr: list[int]) -> list[int]:
    head = _build_list(arr)

    prev = None
    current = head
    while current is not None:
        next_node = current.next
        current.next = prev
        prev = current
        current = next_node

    return _to_array(prev)

File: algorithms/data-structures/linked-list-operations/rust/reverse_linked_list.rs
type Link = Option<Box<Node>>;

struct Node {
    value: i32,
    next: Link,
}

fn build_list(arr: &[i32]) -> Link {
    let mut head: Link = None;
    for &val in arr.iter().rev() {
        head = Some(Box::new(Node {
            value: val,
            next: head,
        }));
    }
    head
}

fn to_array(head: &Link) -> Vec<i32> {
    let mut result = Vec::new();
    let mut current = head;
    while let Some(node) = current {
        result.push(node.value);
        current = &node.next;
    }
    result
}

pub fn reverse_linked_list(arr: &[i32]) -> Vec<i32> {
    let head = build_list(arr);

    let mut prev: Link = None;
    let mut current = head;
    while let Some(mut node) = current {
        current = node.next.take();
        node.next = prev;
        prev = Some(node);
    }

    to_array(&prev)
}
